pkg/manager/member/tikv_upgrader.go

```diff
@@ -14,26 +14,154 @@
 package member
 
 import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
 	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
 	apps "k8s.io/api/apps/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
 )
 
-type tikvUpgrader struct{}
+const (
+	EvictLeaderBeginTime = "evictLeaderBeginTime"
+	EvictLeaderTimeOut   = 3 * time.Minute
+)
+
+type tikvUpgrader struct {
+	pdControl  controller.PDControlInterface
+	podControl controller.PodControlInterface
+	podLister  corelisters.PodLister
+}
 
 // NewTiKVUpgrader returns a tikv Upgrader
-func NewTiKVUpgrader() Upgrader {
-	return &tikvUpgrader{}
+func NewTiKVUpgrader(pdControl controller.PDControlInterface,
+	podControl controller.PodControlInterface,
+	podLister corelisters.PodLister) Upgrader {
+	return &tikvUpgrader{
+		pdControl:  pdControl,
+		podControl: podControl,
+		podLister:  podLister,
+	}
 }
 
 func (tku *tikvUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
 	if tc.Status.PD.Phase == v1alpha1.UpgradePhase {
 		_, podSpec, err := GetLastAppliedConfig(oldSet)
 		if err != nil {
 			return err
 		}
 		newSet.Spec.Template.Spec = *podSpec
-	} else {
-		tc.Status.TiKV.Phase = v1alpha1.UpgradePhase
+		return nil
+	}
+
+	if !tc.Status.TiKV.Synced {
+		return fmt.Errorf("Tidbcluster: [%s/%s]'s tikv status sync failed,can not to be upgraded", ns, tcName)
+	}
+
+	tc.Status.TiKV.Phase = v1alpha1.UpgradePhase
+	setUpgradePartition(newSet, *oldSet.Spec.UpdateStrategy.RollingUpdate.Partition)
+
+	if tc.Status.TiKV.StatefulSet.CurrentReplicas == 0 {
+		return controller.RequeueErrorf("tidbcluster: [%s/%s]'s tikv doesn't have old version pod to upgrade", ns, tcName)
+	}
+
+	if !tc.TiKVAllPodsStarted() {
+		return controller.RequeueErrorf("tidbcluster: [%s/%s]'s tikv pods are not all created", ns, tcName)
+	}
+	for i := tc.Status.TiKV.StatefulSet.Replicas; i > tc.Status.TiKV.StatefulSet.CurrentReplicas; i-- {
+		store := tku.getStoreByOrdinal(tc, i-1)
+		if store == nil || store.State != v1alpha1.TiKVStateUp {
+			return controller.RequeueErrorf("tidbcluster: [%s/%s]'s upgraded tikv pods are not all ready", ns, tcName)
+		}
+	}
+
+	upgradeOrdinal := tc.Status.TiKV.StatefulSet.CurrentReplicas - 1
+	upgradePodName := tikvPodName(tc, upgradeOrdinal)
+	upgradePod, err := tku.podLister.Pods(ns).Get(upgradePodName)
+	if err != nil {
+		return err
+	}
+
+	for _, store := range tc.Status.TiKV.Stores {
+		if store.PodName == upgradePodName {
+			storeID, err := strconv.ParseUint(store.ID, 10, 64)
+			if err != nil {
+				return err
+			}
+			_, evicting := upgradePod.Annotations[EvictLeaderBeginTime]
+
+			if tku.readyToUpgrade(upgradePod, store) {
+				err = tku.pdControl.GetPDClient(tc).EndEvictLeader(storeID)
```

Comment:
The upgrade will not take effect immediately, the `setUpgradePartition` will update StatefulSet only after the end of the current loop. Before the pod is upgraded, the leaders may be migrated to the pod again.